File: Python/Aula 03/Complementares/ex5.py
class Gerenciador:
    lista_processos = []

    @classmethod
    def registrar(cls, processo):
        cls.lista_processos.append(processo)

    @classmethod
    def remover(cls, processo):
        if processo in cls.lista_processos:
            cls.lista_processos.remove(processo)

    @classmethod
    def listar_ativos(cls):
        nomes = [p.nome for p in cls.lista_processos]
        print(f"Processos Ativos: {nomes}")

class Processo:
    def __init__(self, nome):
        self.nome = nome
        Gerenciador.registrar(self)

    def __del__(self):
        Gerenciador.remover(self)

def rotina_temporaria():
    p2 = Processo("PROCESSO_TEMP_2")
    p3 = Processo("PROCESSO_TEMP_3")
    print("Dentro da rotina:")
    Gerenciador.listar_ativos()
    print("Encerrando rotina...")

if __name__ == "__main__":
    p1 = Processo("PROCESSO_MAIN_1")
    
    rotina_temporaria()
    
    print("Fora da rotina:")
    Gerenciador.listar_ativos()

File: C++/Aula 03/Complementares/ex5.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Processo;

class Gerenciador{
    private:
        vector<Processo *> ativos;

        Gerenciador() = default;

    public:
        static Gerenciador &instancia(){
            static Gerenciador g;
            return g;
        }

        void registrar(Processo *p){
            ativos.push_back(p);
        }

        void remover(Processo *p){
            ativos.erase(
                remove(ativos.begin(), ativos.end(), p),
                ativos.end());
        }

        void listar_ativos() const;

        size_t quantidade() const { return ativos.size(); }
};


class Processo{
    private:
        string nome;
        int pid;
        static int proximoPid;

    public:
        Processo(const string &nome) : nome(nome), pid(proximoPid++){
            Gerenciador::instancia().registrar(this);
            cout << "[+] Processo criado:   " << pid << " (" << nome << ")\n";
        }

        ~Processo(){
            Gerenciador::instancia().remover(this);
            cout << "[-] Processo destruído: " << pid << " (" << nome << ")\n";
        }

        Processo(const Processo &) = delete;
        Processo &operator=(const Processo &) = delete;

        int getPid() const { return pid; }
        string getNome() const { return nome; }
};

int Processo::proximoPid = 100;

void Gerenciador::listar_ativos() const{
    cout << "\n  ┌─ Processos ativos (" << ativos.size() << ") ─────────\n";
    if (ativos.empty()){
        cout << "  │  (nenhum)\n";
    }
    else{
        for (const auto *p : ativos)
            cout << "  │  PID " << p->getPid()
                      << "  —  " << p->getNome() << "\n";
    }
    cout << "  └──────────────────────────────\n\n";
}

int main(){
    auto &G = Gerenciador::instancia();

    cout << "=== Escopo externo ===\n";
    Processo p1("init");
    Processo p2("systemd");

    G.listar_ativos();

    {
        cout << "  === Escopo interno A ===\n";
        Processo p3("nginx");
        Processo p4("postgres");

        G.listar_ativos();

        {
            cout << "    === Escopo interno B ===\n";
            Processo p5("worker-1");
            Processo p6("worker-2");

            G.listar_ativos();

            cout << "    [Saindo do escopo B]\n";
        }

        G.listar_ativos();
        cout << "  [Saindo do escopo A]\n";
    }

    G.listar_ativos();
    cout << "[Saindo do main — p1 e p2 serão destruídos]\n";

    return 0;
}